public/src/components/channelManagement/createTestDialog.tsx

```diff
@@ -0,0 +1,168 @@
+import React, { useEffect } from "react";
+import {
+  Button,
+  Dialog,
+  DialogActions,
+  DialogContent,
+  DialogTitle,
+  createStyles,
+  IconButton,
+  TextField,
+  Theme,
+  WithStyles,
+  withStyles,
+} from "@material-ui/core";
+import CloseIcon from "@material-ui/icons/Close";
+
+import useValidatableField from "../../hooks/useValidatableField";
+
+import {
+  getInvalidCharactersError,
+  getEmptyError,
+  createGetDuplicateError,
+} from "./helpers/validation";
+
+const styles = ({}: Theme) =>
+  createStyles({
+    dialogHeader: {
+      display: "flex",
+      alignItems: "center",
+      justifyContent: "space-between",
+      paddingRight: "8px",
+    },
+  });
+
+const NAME_DEFAULT_HELPER_TEXT = "Date format: YYYY-MM-DD_TEST_NAME";
+const NICKNAME_DEFAULT_HELPER_TEXT =
+  "Pick a name for your test that's easy to recognise";
+
+type Mode = "NEW" | "COPY";
+
+interface CreateTestDialogProps {
+  isOpen: boolean;
+  close: () => void;
+  existingNames: string[];
+  existingNicknames: string[];
+  mode: Mode;
+  copiedTestName?: string;
+  copiedTestNickname?: string;
+  createTest: (name: string, nickname: string) => void;
+}
+
+const CreateTestDialog = ({
+  classes,
+  isOpen,
+  close,
+  existingNames,
+  existingNicknames,
+  mode,
+  copiedTestName,
+  copiedTestNickname,
+  createTest,
+}: CreateTestDialogProps & WithStyles<typeof styles>) => {
+  const getDuplicateNameError = createGetDuplicateError(existingNames);
+  const getNameError = (value: string) =>
+    getInvalidCharactersError(value) ||
+    getEmptyError(value) ||
+    getDuplicateNameError(value);
+
+  const [
+    name,
+    setName,
+    nameHasError,
+    nameHelperText,
+    checkName,
+  ] = useValidatableField(NAME_DEFAULT_HELPER_TEXT, getNameError);
+
+  useEffect(() => {
+    if (mode === "COPY" && copiedTestName) {
+      setName(`Copy of ${copiedTestName}`);
+    }
+  }, []);
+
+  const getDuplicateNicknameError = createGetDuplicateError(existingNicknames);
+  const getNicknameError = (value: string) =>
+    getEmptyError(value) || getDuplicateNicknameError(value);
+
+  const [
```

Comment:
All the state returned by `useValidatableField` appears to be used only by a `TextField`.
Why not have a generic text field component with all the state inside? I think this is pretty much what `EditableTextField` does, though as a class component.

Reply:
Will investigate this now!

Reply:
After IRL chat: will revisit this and put in the time to get a nice, general purpose solution